### PR TITLE
Add initial hipSYCL support

### PR DESCRIPTION
**Pull request by illuhad**
This allows the SYCL benchmarks in PRK to be compiled with hipSYCL. Included changes:
* Like triSYCL, hipSYCL does not advertise support for Khronos fp64 or spir extensions, since it does not use OpenCL. So, on hipSYCL, tests should be carried out even if these extensions are not present.
* Like triSYCL, hipSYCL does not need prebuilding kernels
* Depending on the hipSYCL compilation target, CPU and GPU backends may not be available, which may then cause the benchmark to abort since the device selector throws an exception (because no matching device could be found). This is fixed by introducing the `SYCL_TRY_CPU_QUEUE` and `SYCL_TRY_GPU_QUEUE` macros.
The cpu/gpu portions of the benchmarks are only enabled if these macros are set. If not on hipSYCL both are enabled; on hipSYCL they are only enabled if the right backend is targeted (cpu/gpu).
Since the CPU and GPU benchmark runs are already surrounded with `if(1) { ... }` or `if(0) {... }` statements, we could also change this to `if(SYCL_TRY_CPU_QUEUE) {... }` instead of the additional preprocessor if. I did note make this change because the intent behind those if statements was not entirely clear to me.
* A small workaround around a problem with clang's HIP/CUDA toolchain not having `std::abort()` available when compiling for device.

I have tested all SYCL benchmarks both with the hipSYCL CPU and CUDA backend and they work fine, although there seems to be a cleanup issue in hipSYCL on CPU, potentially leading to a crash when the queue goes out of scope. This does not affect the generation and validation of results.

### CODE REVIEW

**Comment by illuhad** [comment]
I believe the Travis build failures are unrelated to this PR.

**Comment by jeffhammond** [comment]
Thanks.  This looks good to me.  I need to refactor the device selection launch stuff, but I'll do that after this is merged.

**Comment by jeffhammond** [comment]
Yeah, SYCL is failing in Travis because of unrelated issues.  MacOS programming environment started choking on triSYCL a while ago and I haven't had time to work on it.

**Comment by keryell** [comment]
I am slow on this MacOS port too, even if I have a Mac Mini on my desk now... :-(

File: Cxx11/prk_util.h [hunk 1]
```diff
@@ -82,13 +82,34 @@
...
+// device selector, it will throw an exception.
+// These macros can be used to check if there's any chance
+// of an implementation targeting a CPU and GPU.
+#if !defined(__HIPSYCL__) || defined(HIPSYCL_PLATFORM_CPU)
+#define SYCL_TRY_CPU_QUEUE 1
+#else
+#define SYCL_TRY_CPU_QUEUE 0
+#endif
+
+#if !defined(__HIPSYCL__) || !defined(HIPSYCL_PLATFORM_CPU)
```
[the remark this answers is not in the document]
Comment (jeffhammond) [reply Cxx11/prk_util.h]: I fixed this.

[the remark this answers is not in the document]
Comment (illuhad) [reply Cxx11/prk_util.h]: Actually, I think it was correct.. There is no `HIPSYCL_PLATFORM_GPU`, there is `HIPSYCL_PLATFORM_CPU`, `HIPSYCL_PLATFORM_CUDA` and `HIPSYCL_PLATFORM_HCC`. What we want to express here is that GPU queues should be tried
* if we are not on hipSYCL
* or if we are on a hipSYCL GPU backend, i.e. `defined(HIPSYCL_PLATFORM_CUDA) || defined(HIPSYCL_PLATFORM_HCC)`. To shorten this, I used the equivalent `!defined(HIPSYCL_PLATFORM_CPU)`.

In other words, GPU queue should be tried if we are either not on hipSYCL or, if we are not on the hipSYCL CPU backend.

[the remark this answers is not in the document]
Comment (jeffhammond) [reply Cxx11/prk_util.h]: Ah ok, I will fix it, perhaps in a more pedantic way.
